burstphoto/align.swift

```diff
@@ -40,23 +41,509 @@ let mfl = device.makeDefaultLibrary()!
+    
+    // measure execution time
+    let t0 = DispatchTime.now().uptimeNanoseconds
+    var t = t0
+    
+    // check that all images are of the same extension
+    let image_extension = image_urls[0].pathExtension
+    let all_extensions_same = image_urls.allSatisfy{$0.pathExtension == image_extension}
+    if !all_extensions_same {throw AlignmentError.inconsistent_extensions}
+    
+    // check that 2+ images were provided
+    let n_images = image_urls.count
+    if n_images < 2 {throw AlignmentError.less_than_two_images}
+    
+    // create output directory
+    let out_dir = NSHomeDirectory() + "/Pictures/Burst Photo/"
+    if !FileManager.default.fileExists(atPath: out_dir) {
+        try FileManager.default.createDirectory(atPath: out_dir, withIntermediateDirectories: true, attributes: nil)
+    }
+    
+    // create a directory for temporary dngs inside the output directory
+    let tmp_dir = out_dir + ".dngs/"
+    try FileManager.default.createDirectory(atPath: tmp_dir, withIntermediateDirectories: true)
+       
+    // ensure that all files are .dng, converting them if necessary
+    var dng_urls = image_urls
+    let convert_to_dng = image_extension.lowercased() != "dng"
+    DispatchQueue.main.async { progress.includes_conversion = convert_to_dng }
+    let dng_converter_path = "/Applications/Adobe DNG Converter.app"
+    let final_dng_conversion = FileManager.default.fileExists(atPath: dng_converter_path)
+    
+    if convert_to_dng {
+        // check if dng converter is installed
+        if !FileManager.default.fileExists(atPath: dng_converter_path) {
+            // if dng coverter is not installed, prompt user
+            throw AlignmentError.missing_dng_converter
+        } else {
+            // the dng converter is installed -> use it
+            dng_urls = try convert_images_to_dng(image_urls, dng_converter_path, tmp_dir)
+            print("Time to convert images: ", Float(DispatchTime.now().uptimeNanoseconds - t) / 1_000_000_000)
+            DispatchQueue.main.async { progress.int += 10000000 }
+            t = DispatchTime.now().uptimeNanoseconds
+        }
+    }
+       
+    // load images
+    t = DispatchTime.now().uptimeNanoseconds
+    var (textures, mosaic_pattern_width, white_level, black_level, exposure_bias) = try load_images(dng_urls)
+    print("Time to load all images: ", Float(DispatchTime.now().uptimeNanoseconds - t) / 1_000_000_000)
+    t = DispatchTime.now().uptimeNanoseconds
+    DispatchQueue.main.async { progress.int += (convert_to_dng ? 10000000 : 20000000) }
+    
+    // if images have a uniform exposure: use central image as reference
+    var ref_idx = image_urls.count / 2
+       
+    // if images have different exposures: use image with lowest exposure as reference to protect highlights
```

Comment:
I don't think this is a good choice. If I shoot a 5 burst stack at +/-4, +/-2, and 0, The -4 shot will be used as the reference shot, leading to an incredibly dark scene.

A bit more analysis of the images may be required.

Reply:
I believe that this is the only way to protect the highlights, which is typically the main reason to underexpose. Otherwise, you could also use a bracket of 0EV, +1EV, +2EV, ... When you activate the tone-mapping and have an underexposed reference image, your image will look very similar to the 0 EV image, but with a nice roll-off at the highlights.

Reply:
Let me start by describing how I shoot these brackets.

Unless I'm on a tripod and can dial the exposure in slowly, I'll use my camera's exposure burst. So it will shoot 2N+1 images with a Y EV difference(so 1 with an adjustment of 0 EV and then N photos on either side, each with a Y EV difference between them). I will usually do something like 5 photos with 2 EV difference. hence (-4, -2, 0, +2, +4). When I shoot like this, I can only see the 0EV on the screen. So I will shoot the 0EV with "correct" exposure, where I'll accept some highlight clipping in order to get a safe image that I can still use.

This means that my lowest exposure image will be mostly very dark, which will expose properly for any really bright highlights.

When I put one of these burst through the app, the end result looks worse than the single image.

I don't know how many cameras provide a one sided exposure bracket and how many have a symmetric burst (or one at all).

I think perhaps it may be worth using the median value if the app recognizes a symmetric burst. Or maybe one lower than that (in order to preserve any highlights) as a rule of thumb so that the image does not have to get analyzed.

Reply:
On my camera, I can use the exposure compensation dial to adjust the bracketing that it becomes asymmetric, e.g. from -1EV, 0EV, +1EV to 0EV, +1EV, +2EV...
Furthermore I would recommend to start with smaller steps between the bracketing. From -4EV to +4EV, you have 8 stops in between. That is really a lot! I think in most situations it is not possible to hand-hold the +4 EV image. You may try what happens when you use -2EV to +2EV. You will still get a considerable improvement of dynamic range. 2 EV in the highlights and 1-2 EVs in the shadows. 
It is correct, that the merging does not work anymore when the image with the lowest exposure is completely black. In that case, the alignment does not work anymore. So it is better to use a smaller bracket as long as you are not on a tripod and can apply the simple temporal averaging.

Reply:
Topic resolved in offline discussion.